interactions/ext/hybrid_commands/hybrid_slash.py

```diff
@@ -0,0 +1,593 @@
+        elif lowered in {"no", "n", "false", "f", "0", "disable", "off"}:  # noqa: RET505
+            return False
+        raise BadArgument(f"{argument} is not a recognised boolean option.")
+
+
+class AttachmentConverter(NoArgumentConverter):
+    async def convert(self, ctx: "HybridContext", _: Any) -> Attachment:
+        try:
+            attachment = ctx.message.attachments[ctx.__attachment_index__]
+            ctx.__attachment_index__ += 1
+            return attachment
+        except IndexError:
+            raise BadArgument("No attachment found.") from None
+
+
+class ChoicesConverter(_LiteralConverter):
+    def __init__(self, choices: list[SlashCommandChoice | dict]) -> None:
+        standardized_choices = tuple((SlashCommandChoice(**o) if isinstance(o, dict) else o) for o in choices)
+
+        names = tuple(c.name for c in standardized_choices)
+        self.values = {str(arg): str for arg in names}
+        self.choice_values = {str(c.name): c.value for c in standardized_choices}
+
+    async def convert(self, ctx: BaseContext, argument: str) -> Any:
+        val = await super().convert(ctx, argument)
+        return self.choice_values[val]
+
+
+class RangeConverter(Converter[float | int]):
+    def __init__(
+        self,
+        number_type: int,
+        min_value: Optional[float | int],
+        max_value: Optional[float | int],
+    ) -> None:
+        self.number_type = number_type
+        self.min_value = min_value
+        self.max_value = max_value
+
+        self.number_convert = int if number_type == OptionType.INTEGER else float
+
+    async def convert(self, ctx: BaseContext, argument: str) -> float | int:
+        try:
+            converted: float | int = await maybe_coroutine(self.number_convert, ctx, argument)
+
+            if self.min_value and converted < self.min_value:
+                raise BadArgument(f'Value "{argument}" is less than {self.min_value}.')
+            if self.max_value and converted > self.max_value:
+                raise BadArgument(f'Value "{argument}" is greater than {self.max_value}.')
+
+            return converted
+        except ValueError:
+            type_name = "number" if self.number_type == OptionType.NUMBER else "integer"
+
+            if type_name.startswith("i"):
+                raise BadArgument(f'Argument "{argument}" is not an {type_name}.') from None
+            raise BadArgument(f'Argument "{argument}" is not a {type_name}.') from None
+        except BadArgument:
+            raise
+
+
+class StringLengthConverter(Converter[str]):
+    def __init__(self, min_length: Optional[int], max_length: Optional[int]) -> None:
+        self.min_length = min_length
+        self.max_length = max_length
+
+    async def convert(self, ctx: BaseContext, argument: str) -> str:
+        if self.min_length and len(argument) < self.min_length:
+            raise BadArgument(f'The string "{argument}" is shorter than' f" {self.min_length} character(s).")
```

Comment:
Why not keep it to one f-string?

```suggestion
            raise BadArgument(f'The string "{argument}" is shorter than {self.min_length} character(s).')
```

Reply:
weird black moment